src/smtencoding/FStar.SMTEncoding.Encode.fs

```diff
@@ -1481,45 +1482,95 @@ and encode_sigelt' (env:env_t) (se:sigelt) : (decls_t * env_t) =
-                                              mkImp(ty_pred, mk_and_l prec)),
-                                    Some "subterm ordering",
-                                    ("subterm_ordering_"^ddconstrsym))
                   in
-                  arg_decls, [typing_inversion; subterm_ordering]
-
+                  Util.mkAssume(mkForall (Ident.range_of_lid d)
+                                         ([[ty_pred]],
+                                          add_fuel (mk_fv (fuel_var, Fuel_sort)) (vars@arg_binders),
+                                          mkImp(ty_pred, mk_and_l prec)),
+                                Some "subterm ordering",
+                                ("subterm_ordering_"^ddconstrsym))
+              in
+              let codomain_ordering =
+                let tot_or_gtot_inductive_codomain c =
+                  let is_inductive l =
+                      match FStar.TypeChecker.Env.lookup_sigelt env.tcenv l with
+                      | Some ({sigel=Sig_inductive_typ _}) -> true
+                      | _ -> false
+                  in
+                  let res =
+                   if not (U.is_tot_or_gtot_comp c)
+                   then false
+                   else let head, _ = U.head_and_args (U.comp_result c) in
+                        BU.for_some
+                          (fun mutual -> is_inductive mutual && U.is_fvar mutual head)
+                          mutuals
+                  in
+                  res
+                in
+                let codomain_prec_l =
+                  List.collect
+                    (fun (formal, var) ->
+                        let bs, c = U.arrow_formals_comp formal.binder_bv.sort in
+                        match bs with
+                        | [] -> []
+                        | _ when not (tot_or_gtot_inductive_codomain c) -> []
+                        | _ ->
+                         //var bs << D ... var ...
+                         let bs', guards', env', bs_decls, _ = encode_binders None bs env'' in
+                         let fun_app = mk_Apply (mkFreeV var) bs' in
+                         [mkForall (Ident.range_of_lid d)
+                                   ([[mk_Precedes lex_t lex_t fun_app dapp]],
+                                    bs',
+                                    mkImp (mk_and_l guards',
+                                           mk_Precedes lex_t lex_t fun_app dapp))])
+                    (List.zip formals vars)
+                in
+                match codomain_prec_l with
+                | [] -> []
                 | _ ->
-                  Errors.log_issue se.sigrng (Errors.Warning_ConstructorBuildsUnexpectedType, (BU.format2 "Constructor %s builds an unexpected type %s\n"
-                        (Print.lid_to_string d) (Print.term_to_string head)));
-                  [], [] in
+                  [Util.mkAssume(mkForall (Ident.range_of_lid d)
```

Comment:
We should perhaps increment the checked file version number since smt encoding is also cached?